### PR TITLE
ECL-462: Amending A Return Journey - Create PDF and Send to DMS

### PR DESCRIPTION
**Pull request by nemanjatomanovic**

### CODE REVIEW

File: app/uk/gov/hmrc/economiccrimelevyreturns/services/DmsService.scala [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+)(implicit
+  ec: ExecutionContext
+) {
+
+  def submitToDms(optBase64EncodedDmsSubmissionHtml: Option[String], now: Instant)(implicit
+    hc: HeaderCarrier
+  ): Future[Either[UpstreamErrorResponse, SubmitEclReturnResponse]] =
+    optBase64EncodedDmsSubmissionHtml match {
+      case Some(base64EncodedDmsSubmissionHtml) =>
+        val html = new String(Base64.getDecoder.decode(base64EncodedDmsSubmissionHtml))
```
Comment (liandra-giacomine) [inline app/uk/gov/hmrc/economiccrimelevyreturns/services/DmsService.scala:49]: nitpick - .map(_.toChar).mkString is more functional

File: app/uk/gov/hmrc/economiccrimelevyreturns/models/Done.scala [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.economiccrimelevyreturns.models
+
+sealed abstract class Done
```
[the remark this answers is not in the document]
Comment (nemanjatomanovic) [reply app/uk/gov/hmrc/economiccrimelevyreturns/models/Done.scala:19]: `Done` is part of the pattern that is introduced in registration service that helps us to identify if core parts of Internal auth token initialisation process are Done. I believe that it has a value in this process, and not only 'ceremonial' one, being part of the pattern.

File: app/uk/gov/hmrc/economiccrimelevyreturns/config/InternalAuthTokenInitialiser.scala [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+
+    Future
+      .sequence(seq)
+      .map(_ => Done)
+  }
+
+  Await.result(initialised, 30.seconds)
+
+  private def ensureAuthToken(): Future[Done] =
+    authTokenIsValid.flatMap { isValid =>
```
[the remark this answers is not in the document]
Comment (nemanjatomanovic) [reply app/uk/gov/hmrc/economiccrimelevyreturns/config/InternalAuthTokenInitialiser.scala]: Addressed.

File: app/uk/gov/hmrc/economiccrimelevyreturns/connectors/DmsConnector.scala [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+  )(implicit hc: HeaderCarrier): Future[Either[UpstreamErrorResponse, HttpResponse]] =
+    retryFor[Either[UpstreamErrorResponse, HttpResponse]]("DMS submission")(retryCondition) {
+      httpClient
+        .post(new URL(appConfig.dmsSubmissionUrl))
+        .setHeader(AUTHORIZATION -> appConfig.internalAuthToken)
+        .withBody(body)
+        .execute[Either[UpstreamErrorResponse, HttpResponse]]
+        .map {
+          case Right(httpResponse: HttpResponse)          => Right(httpResponse)
+          case Left(errorResponse: UpstreamErrorResponse) => throw errorResponse
```
[the remark this answers is not in the document]
Comment (nemanjatomanovic) [reply app/uk/gov/hmrc/economiccrimelevyreturns/connectors/DmsConnector.scala]: Addressed.

File: app/uk/gov/hmrc/economiccrimelevyreturns/controllers/ReturnSubmissionController.scala [hunk 5]
```diff
@@ -44,16 +49,29 @@ class ReturnSubmissionController @Inject() (
...
+                  response =>
+                    nrsService.submitToNrs(
+                      eclReturn.base64EncodedNrsSubmissionHtml,
+                      request.eclRegistrationReference
+                    )
+
+                    Ok(Json.toJson(response))
+                }
+              case AmendReturn     =>
+                val now = Instant.now
```
[the remark this answers is not in the document]
Comment (nemanjatomanovic) [reply app/uk/gov/hmrc/economiccrimelevyreturns/controllers/ReturnSubmissionController.scala]: Addressed.

File: app/uk/gov/hmrc/economiccrimelevyreturns/services/DmsService.scala (hunk 1 above)
Comment (nemanjatomanovic) [reply app/uk/gov/hmrc/economiccrimelevyreturns/services/DmsService.scala]: Addressed.

File: app/uk/gov/hmrc/economiccrimelevyreturns/controllers/DmsNotificationController.scala [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+      resourceLocation = ResourceLocation("dms-returns-callback")
+    ),
+    action = IAAction("WRITE")
+  )
+
+  private val authorised = auth.authorizedAction(predicate)
+
+  def dmsCallback: Action[JsValue] = authorised(parse.json) { implicit request =>
+    request.body.validate[DmsNotification] match {
+      case JsSuccess(notification, _) =>
```
[the remark this answers is not in the document]
Comment (nemanjatomanovic) [reply app/uk/gov/hmrc/economiccrimelevyreturns/controllers/DmsNotificationController.scala]: Addressed.

File: app/uk/gov/hmrc/economiccrimelevyreturns/config/InternalAuthTokenInitialiser.scala [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+    val resources = ensureResourceDefinitions()
+    val grants    = ensureGrants()
+    val seq       = Seq(authToken, resources, grants)
+
+    Future
+      .sequence(seq)
+      .map(_ => Done)
+  }
+
+  Await.result(initialised, 30.seconds)
```
[the remark this answers is not in the document]
Comment (liandra-giacomine) [reply app/uk/gov/hmrc/economiccrimelevyreturns/config/InternalAuthTokenInitialiser.scala:59]: To make this functional we can make use of the cats library and use IO instead

**Review by liandra-giacomine — APPROVED** [review]

**Pull request merged**